### PR TITLE
Add support for select multiple request.

### PR DESCRIPTION
**Pull request by rohitjakhar**
## :page_facing_up: Context
Issue link: [Share Multiple Request at One Time](https://github.com/ChuckerTeam/chucker/issues/1127)

## :pencil: Changes

Made change in TransactionAdapter to show color on Selected. Transaction, list of selected position, lambda for longClick on itemView.
Also Add query for get selected transaction data, delete transaction data.

## :hammer_and_wrench: How to test
Long press at any transaction to select transaction, after that click other transaction to select them too.

### CODE REVIEW

**Comment by rohitjakhar** [comment]
@cortinico what should i do next?

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainActivity.kt [hunk 1]
```diff
@@ -66,9 +67,26 @@ internal class MainActivity :
...
-        }
+        transactionsAdapter =
+            TransactionAdapter(
+                context = this,
+                longPress = { transactionId ->
+                    selectedTransactions.add(transactionId)
+                },
+                onTransactionClick = { transactionId ->
+                    if (selectedTransactions.isNotEmpty()) {
+                        if (selectedTransactions.contains(transactionId)) {
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainActivity.kt:78]:
```suggestion
                        if (transactionId in selectedTransactions) {
```

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainActivity.kt [hunk 2]
```diff
@@ -148,15 +168,16 @@ internal class MainActivity :
                 showDialog(
                     getClearDialogData(),
                     onPositiveClick = {
-                        viewModel.clearTransactions()
+                        if (selectedTransactions.isNotEmpty()) viewModel.clearSelectedTransactions(selectedTransactions) else viewModel.clearTransactions()
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainActivity.kt:171]: Can we move this logic inside hte ViewModel? Let's just have a `viewModel.clearTransactions(selectedTransactions)` that behaves differently wether the `selectedTransactions` is empty or not

File: library/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -47,7 +47,10 @@
     <string name="chucker_share_transaction_subject">Transaction details</string>
     <string name="chucker_share_all_transactions_subject">All transactions</string>
     <string name="chucker_clear_http_confirmation">Do you want to clear complete network calls history?</string>
+    <string name="chucker_clear_selected_http_confirmation">Do you want to clear selected network calls history?</string>
```
Comment (cortinico) [inline library/src/main/res/values/strings.xml:50]:
```suggestion
    <string name="chucker_clear_selected_http_confirmation">Do you want to clear the selected network transactions?</string>
```

File: library/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -47,7 +47,10 @@
     <string name="chucker_share_transaction_subject">Transaction details</string>
     <string name="chucker_share_all_transactions_subject">All transactions</string>
     <string name="chucker_clear_http_confirmation">Do you want to clear complete network calls history?</string>
```
Comment (cortinico) [inline library/src/main/res/values/strings.xml:49]:
```suggestion
    <string name="chucker_clear_http_confirmation">Do you want to clear the complete network calls history?</string>
```

**Review by cortinico — CHANGES_REQUESTED** [review]
Thanks for sending this over. I've done a first pass. We need to do more testing as currently the logic is broken, see this video:

untitled

**Comment by rohitjakhar** [comment]
Thanks @cortinico  for review, i will try to fix that ASAP.

**Comment by cortinico** [comment]
@rohitjakhar Just to set expectations, I'll be able to take a look at this only next week.

**Comment by rohitjakhar** [comment]
@cortinico  I resolve conflicts, test all corner cases.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionAdapter.kt [hunk 5]
```diff
@@ -63,14 +84,39 @@ internal class TransactionAdapter internal constructor(
...
+                } ?: kotlin.run { false }
+            }
         }
 
         @SuppressLint("SetTextI18n")
         fun bind(transaction: HttpTransactionTuple) {
             transactionId = transaction.id
-
+            if (selectedPos.find { it == adapterPosition } != null) {
+                itemView.setBackgroundColor(color400)
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionAdapter.kt:116]: Let's not used this.
Let's create a new color resource as:
```xml
<color name="chucker_status_highlighted">#FF9800</color>
```

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionAdapter.kt [hunk 6]
```diff
@@ -63,14 +84,39 @@ internal class TransactionAdapter internal constructor(
...
+                transactionId?.let {
+                    longPress.invoke(it)
+                    if (selectedPos.contains(adapterPosition)) {
+                        selectedPos.remove(adapterPosition)
+                    } else {
+                        selectedPos.add(adapterPosition)
+                    }
+                    notifyItemChanged(adapterPosition)
+                    true
+                } ?: kotlin.run { false }
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionAdapter.kt:108]:
```suggestion
                } ?: false
```

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionAdapter.kt [hunk 7]
```diff
@@ -35,6 +40,22 @@ internal class TransactionAdapter internal constructor(
...
     private val color400: Int = ContextCompat.getColor(context, R.color.chucker_status_400)
     private val color300: Int = ContextCompat.getColor(context, R.color.chucker_status_300)
+    val outValue = TypedValue()
+    @Suppress("UnusedPrivateProperty")
+    private val defaultColor =
+        context.theme.resolveAttribute(
+        android.R.attr.selectableItemBackground,
+        outValue,
+        true,
+    )
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionAdapter.kt:50]: formatting is odd also here

**Review by cortinico — CHANGES_REQUESTED** [review]
Thanks again for sending this over @rohitjakhar 

Please reformat the code with `./gradlew ktlintFormat` as formatting is odd in a couple of places.

Also there is a problem with rotation of the device: the highlights is lost, but the selection still remains. Also clicking delete is deleting the wrong transaction. See the video I'm attaching.

untitled

**Comment by cortinico** [comment]
We're getting close to a mergeabble state :)

**Comment by rohitjakhar** [comment]
@cortinico I fix above issues.

**Comment by rohitjakhar** [comment]
@cortinico can you give me next step to merge it?

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainActivity.kt [hunk 8]
```diff
@@ -150,14 +164,21 @@ internal class MainActivity :
...
             R.id.share_text -> {
                 showDialog(
-                    getExportDialogData(R.string.chucker_export_text_http_confirmation),
+                    getExportDialogData(
+                        if (viewModel.isItemSelected.value == true) {
+                            R.string.chucker_export_text_selected_http_confirmation
+                        } else {
+                            R.string.chucker_export_text_http_confirmation
+                        },
+                    ),
```
Comment (cortinico) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainActivity.kt:181]: Because of this, the function is now getting too long and letting detekt fail:

```

> Task :library:detekt
/home/runner/work/chucker/chucker/library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainActivity.kt:161:18: The function onOptionsItemSelected is too long (60). The maximum length is 60. [LongMethod]

```

You can see it in the CI signal.

Can you export this + lines 193/199 to separate functions

**Review by cortinico — CHANGES_REQUESTED** [review]
Let's make sure the CI is green 👍 You also need to run `./gradlew ktlintFormat` before resubmitting

**Comment by rohitjakhar** [comment]
@cortinico  I just update code and run ktlintformat.

**Comment by rohitjakhar** [comment]
@cortinico  should i supress TooManyFunctions ?

**Comment by cortinico** [comment]
> @cortinico should i supress TooManyFunctions ?

Yes please do, it's reasonable to do so in `MainActivity.kt`

**Comment by rohitjakhar** [comment]
@cortinico  All are green

**Comment by cortinico** [comment]
@rohitjakhar thanks for following up.

Sadly there are still a couple of bugs:

1. The 4xx transactions are rendered impossible to read (see screnshot). We should probably change the text color to black when things are selected.

Screenshot_1718923015

2. If you quickly select transactions as they're getting populated, you end up in a broken state where not all selected transactions are actually deleted. See video:


https://github.com/ChuckerTeam/chucker/assets/3001957/36f5c85b-c2b8-4f54-ac05-1a4468dd8f61

Can I ask you to look into those 2 problems and then we can merge it 👍

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainViewModel.kt [hunk 9]
```diff
@@ -32,15 +38,38 @@ internal class MainViewModel : ViewModel() {
...
+            } else {
+                selectedItemId.value?.add(itemId)
+                _isItemSelected.value = true
+            }
+        }
+    }
+
+    suspend fun getAllTransactions(): List<HttpTransaction> {
+        return if (isItemSelected.value == true) {
+            RepositoryProvider.transaction().getSelectedTransactions(selectedItemId.value!!)
```
Comment (shivanandyadav11) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainViewModel.kt:55]: Can we add a safe check `?` instead of `!!` ?

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainViewModel.kt [hunk 10]
```diff
@@ -32,15 +38,38 @@ internal class MainViewModel : ViewModel() {
...
     fun updateItemsFilter(searchQuery: String) {
         currentFilter.value = searchQuery
     }
 
     fun clearTransactions() {
         viewModelScope.launch {
-            RepositoryProvider.transaction().deleteAllTransactions()
+            if (isItemSelected.value == true) {
+                _isItemSelected.value = false
+                RepositoryProvider.transaction().deleteSelectedTransactions(selectedItemId.value!!)
```
Comment (shivanandyadav11) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/MainViewModel.kt:69]: Same as above

**Review by shivanandyadav11 — APPROVED** [review]

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionAdapter.kt [hunk 11]
```diff
@@ -35,6 +41,32 @@ internal class TransactionAdapter internal constructor(
...
+    private val defaultColor =
+        context.theme.resolveAttribute(
+            android.R.attr.selectableItemBackground,
+            outValue,
+            true,
+        )
+
+    fun getSelectedItem(): List<Int> {
+        return selectedPos
+    }
```
Comment (shivanandyadav11) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionAdapter.kt:57]:
```suggestion
    fun getSelectedItem(): List<Int> =  selectedPos
```

**Comment by cortinico** [comment]
@rohitjakhar if you can't continue working on this on please let me know 👍 perhaps @shivanandyadav11 you want to take it over?

**Comment by shivanandyadav11** [comment]
> @rohitjakhar if you can't continue working on this on please let me know 👍 perhaps @shivanandyadav11 you want to take it over?

@cortinico Sure, I can pick this up. I'll start working on it this coming weekend.

**Comment by cortinico** [comment]
Let's wait for a couple of days for a response from @rohitjakhar
